File: golang/structure/proxy.go
package main
type IUser interface {
    login(string) bool
}
type User struct {
}
func (u *User) login(name string) bool {
    println(name, "login process")
    return true
}
type UserProxy struct {
    user User
}
func (p *UserProxy) login(name string) bool {
    println(name, "login begin ...")
    res := p.user.login(name)
    println(name, "login end ...")
    return res
}
func main() {
    proxy := UserProxy{User{}}
    proxy.login("tom")
}

File: cpp/structure/proxy.cpp
#include<iostream>
using namespace std;
class IUser {
public:
    virtual bool login(string name) = 0;
};
class User: public IUser {
public:
    bool login(string name) {
        return true;
    }
};
class UserProxy: public IUser {
private:
    User user;
public:
    bool login(string name) {
        cout<<name <<" login ..." <<endl;
        bool result = user.login(name);
        cout<<name <<" login success" <<endl;
        return result;
    }
};


int main() {
    UserProxy proxy;
    proxy.login("tom");
    return 0;
}